unify build finding logic

grosser opened the pull request:
@jonmoter @ragurney 

WIP ... but where I think we need to go

jonmoter commented on lib/samson/build_finder.rb:
```diff
@@ -26,51 +26,55 @@ def cancelled!
+          found = detect_build_by_selector!(possible_builds, *selector, fail: (last_try && build_disabled))
+          if found
+            all << found
+          elsif last_try && !build_disabled
```
Kindof duplicated

jonmoter commented on plugins/kubernetes/app/models/kubernetes/template_filler.rb:
```diff
@@ -69,14 +69,26 @@ def set_secrets
+        [nil, container.fetch(:image)]
+      else
+        selected = container[:"samson/dockerfile"] || 'Dockerfile'
+        next if selected == 'none'
```
This is causing an error. Can't use next when not in a loop.

grosser replied:
just need your time for the review, I can debug travis ;)

ragurney commented on lib/samson/build_finder.rb:
```diff
@@ -26,51 +26,56 @@ def cancelled!
+      all
+    end
+
+    def self.detect_build_by_selector!(builds, dockerfile, image, fail:)
```
Would this be more clear as:
```ruby
    class << self
      def detect_build_by_selector!(builds, dockerfile, image, fail:)
        image_name = image.split('/').last.split(/[:@]/, 2).first if image

        if (matching_build = find_matching_build(builds, dockerfile, image_name))
          matching_build
        elsif fail
          raise(
            Samson::Hooks::UserError,
            "Did not find build for dockerfile #{dockerfile.inspect} or image_name #{image_name.inspect}.\n" \
            "Found builds: #{builds.map { |b| [b.dockerfile, b.image_name] }.uniq.inspect}."
          )
        end
      end

      private

      def find_matching_build(builds, dockerfile, image_name)
        builds.detect do |b|
          (image_name && b.image_name == image_name) || (dockerfile && b.dockerfile == dockerfile)
        end
      end
    end
```
?

grosser commented:
kinda similar ... maybe flattening it our with a early return/raise helps,
I'll play with this a bit

On Wed, Feb 14, 2018 at 10:56 AM, Ryan Gurney <notifications@github.com>
wrote:

> *@ragurney* commented on this pull request.
> ------------------------------
>
> In lib/samson/build_finder.rb
> <https://github.com/zendesk/samson/pull/2573#discussion_r168275218>:
>
> > +        needed.delete_if do |selector|
> +          found = detect_build_by_selector!(possible_builds, *selector, fail: (last_try && build_disabled))
> +          if found
> +            all << found
> +          elsif last_try
> +            raise if build_disabled # should never get here
> +            all << create_build(dockerfile)
> +          end
> +        end
> +        needed.empty? # stop the waiting when we got everything
> +      end
> +
> +      all
> +    end
> +
> +    def self.detect_build_by_selector!(builds, dockerfile, image, fail:)
>
> Would this be more clear as:
>
>     class << self
>       def detect_build_by_selector!(builds, dockerfile, image, fail:)
>         image_name = image.split('/').last.split(/[:@]/, 2).first if image
>
>         if (matching_build = find_matching_build(builds, dockerfile, image_name))
>           matching_build
>         elsif fail
>           raise(
>             Samson::Hooks::UserError,
>             "Did not find build for dockerfile #{dockerfile.inspect} or image_name #{image_name.inspect}.\n" \
>             "Found builds: #{builds.map { |b| [b.dockerfile, b.image_name] }.uniq.inspect}."
>           )
>         end
>       end
>
>       private
>
>       def find_matching_build(builds, dockerfile, image_name)
>         builds.detect do |b|
>           (image_name && b.image_name == image_name) || (dockerfile && b.dockerfile == dockerfile)
>         end
>       end
>     end
>
> ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/2573#pullrequestreview-96616314>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ4YV4CjEbf0CO-bC0IqW8_GkuoDtks5tUyx6gaJpZM4SEWvW>
> .
>

ragurney commented on plugins/kubernetes/app/models/kubernetes/template_filler.rb:
```diff
@@ -69,14 +69,25 @@ def set_secrets
+      else
+        selected = container[:"samson/dockerfile"] || 'Dockerfile'
+        next if selected == 'none'
+        [selected, nil]
```
Would this break if someone has their dockerfile as `Dockerfile` and specifies an image name on their build? Then they would have a build with a [nil docker file](https://github.com/zendesk/samson/blob/master/app/models/build.rb#L76) and we wouldn't be able to find a matching build.

grosser replied:
yes, builds with nil `dockerfile` column cannot be found when not using `external images` feature, but that's a separate discussion/feature ... trying to keep this refactoring only :)

jonmoter commented:
I'd rather focus on getting the fix out and making it look cleaner later. There are devs who are blocked by this bug, and unable to deploy.

grosser commented:
oh yeah ... all green without having breaking any tests, I got high hopes for this one :D
